File: backend/app/models/challenge.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class Challenge(Base):
    """Global challenge definitions, seeded via migration — same shape as
    `Achievement`, but time-boxed (weekly/monthly) rather than lifetime."""

    __tablename__ = "challenges"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    key: Mapped[str] = mapped_column(String(64), unique=True, index=True, nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    metric: Mapped[str] = mapped_column(String(32), nullable=False)  # volume_kg | session_count | streak_days
    period: Mapped[str] = mapped_column(String(16), nullable=False)  # weekly | monthly | ongoing
    target_value: Mapped[float] = mapped_column(Float, nullable=False)
    icon: Mapped[str] = mapped_column(String(32), nullable=False, default="flag")
    is_group: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")


class UserChallenge(Base):
    __tablename__ = "user_challenges"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    challenge_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("challenges.id"), nullable=False, index=True)
    joined_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    period_start: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    progress_value: Mapped[float] = mapped_column(Float, nullable=False, default=0)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    user: Mapped["User"] = relationship(back_populates="challenge_entries")
    challenge: Mapped["Challenge"] = relationship()

File: app/lib/features/challenges/domain/challenge.dart
class Challenge {
  const Challenge({
    required this.id,
    required this.key,
    required this.title,
    required this.description,
    required this.metric,
    required this.period,
    required this.targetValue,
    required this.icon,
    required this.isGroup,
  });

  final String id;
  final String key;
  final String title;
  final String description;
  final String metric; // volume_kg | session_count | streak_days
  final String period; // weekly | monthly | ongoing
  final double targetValue;
  final String icon;
  final bool isGroup;

  factory Challenge.fromJson(Map<String, dynamic> json) => Challenge(
    id: json['id'] as String,
    key: json['key'] as String,
    title: json['title'] as String,
    description: json['description'] as String,
    metric: json['metric'] as String,
    period: json['period'] as String,
    targetValue: (json['target_value'] as num).toDouble(),
    icon: json['icon'] as String,
    isGroup: json['is_group'] as bool,
  );
}

class UserChallenge {
  const UserChallenge({
    required this.challenge,
    required this.joined,
    required this.progressValue,
    this.periodStart,
    this.completedAt,
  });

  final Challenge challenge;
  final bool joined;
  final double progressValue;
  final DateTime? periodStart;
  final DateTime? completedAt;

  bool get isCompleted => completedAt != null;
  double get progressFraction => challenge.targetValue <= 0 ? 0 : (progressValue / challenge.targetValue).clamp(0, 1);

  factory UserChallenge.fromJson(Map<String, dynamic> json) => UserChallenge(
    challenge: Challenge.fromJson(json['challenge'] as Map<String, dynamic>),
    joined: json['joined'] as bool,
    progressValue: (json['progress_value'] as num).toDouble(),
    periodStart: json['period_start'] != null ? DateTime.parse(json['period_start'] as String) : null,
    completedAt: json['completed_at'] != null ? DateTime.parse(json['completed_at'] as String) : null,
  );
}

class LeaderboardEntry {
  const LeaderboardEntry({required this.rank, required this.displayName, required this.progressValue, required this.isMe});

  final int rank;
  final String displayName;
  final double progressValue;
  final bool isMe;

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) => LeaderboardEntry(
    rank: json['rank'] as int,
    displayName: json['display_name'] as String,
    progressValue: (json['progress_value'] as num).toDouble(),
    isMe: json['is_me'] as bool,
  );
}
